internal/credentials/kubernetes/acr/workload_identity.go

```diff
@@ -0,0 +1,330 @@
+package acr
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/AzureAD/microsoft-authentication-library-for-go/apps/confidential"
+	"github.com/patrickmn/go-cache"
+	v1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/credentials"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	accessTokenUsername  = "00000000-0000-0000-0000-000000000000"
+	metadataURL          = "http://169.254.169.254/metadata/instance?api-version=2021-02-01"
+	viewerRoleName       = "kargo-viewer"
+	AzureDefaultAudience = "api://AzureADTokenExchange"
+	AnnotationClientID   = "azure.workload.identity/client-id"
+	AnnotationTenantID   = "azure.workload.identity/tenant-id"
```

Comment:
I think we are best off getting tenantID from values.yaml at install time. Similar for EntraID (nee AAD) endpoint URL. I _think_ we can entirely forget about the metadataURL and the steps to retrieve Azure environment if tenantID and EntraID endpoint URL are explicitly configured at install time.

Reply:
> I think we can entirely forget about the metadataURL and the steps to retrieve Azure environment if tenantID and EntraID endpoint URL are explicitly configured at install time.

~That is true, although I don't know how easy it is for a user to find the EntraID endpoint URL...~ EDIT: Nevermind, the endpoints can be found [here](https://learn.microsoft.com/en-us/entra/identity-platform/authentication-national-cloud)

Reply:
We also need to determine the [management endpoint](https://learn.microsoft.com/en-us/rest/api/azure/). From what I can see it's `https://management.core.windows.net/` when I test this, but I don't know if this will be true for all environments.

Reply:
Ah, these endpoints are cloud specific:

The default Public Cloud has `https://management.core.windows.net`, but the endpoint differ with the other variants (China Cloud, Gov Cloud etc). We could perhaps do what external-secrets does and have the Azure Cloud Type (`AZURECHINACLOUD`, `AZUREPUBLICCLOUD` etc) be set as a Helm value / env variable, and then get those endpoints with a map lookup without needing to call the metadata API?

Reply:
I've been racking my brain, because I used to know an easier way to do this. I finally found it... sort of...

If "environment" is configured at install-time with a well-known value like `AzurePublicCloud` or `AzureChinaCloud`:

```golang
import "github.com/Azure/go-autorest/autorest/azure"

// ...

azureEnv, err := azure.EnvironmentFromName(envName)
// ...
```

The returned `azure.Environment` struct includes _some_ of the same info you are obtaining right now using a direct HTTP call to the Azure Instance Metadata Service. We can get both the AAD endpoint and the resource management endpoint.

In the end, I like your existing approach of querying IMDS directly much better. I'd initially found it objectionable because of the direct HTTP request. It'd have been nice if we could have accomplished everything using official Azure SDKs. Although `azure.EnvironmentFromName(envName)` looks tidier, it comes at the price of requiring more configuration at install-time.

So... as long as we're making that call to IMDS, we _might_ be able to get tenantID from there as well. It's not documented, but I _suspect_ it might be there. I will experiment to try and find out.

We can fall back on getting tenantID from the `AZURE_TENANT_ID` env var like you do currently, _but_ if we're able to get the tenant ID without looking at `AZURE_TENANT_ID`, the advantage will be no need to establish a managed identity for the controller itself. You could bother with managed identity for the Projects only. So, if we _can't_ get tenantID from IMDS, there may still be a benefit in having the user provide tenantID at install-time since it helps eliminate the need for the controller to have a managed identity.

Of course, if we want functionality like we have for ECR where we fall back on the controller's own permissions if the Project doesn't have a managed identity, then the controller _does_ need a managed identity.

I think all of this is going to be one giant game of starting with the ideal where no extra configuration is required at install-time and only Projects need managed identity and then we start gracefully falling back on other strategies as needed.

Reply:
> So... as long as we're making that call to IMDS, we might be able to get tenantID from there as well. It's not documented, but I suspect it might be there. I will experiment to try and find out.

Nope. It's not there. At least not when I `curl -H "Metadata: true" http://169.254.169.254/metadata/instance?api-version=2021-02-01` from within an Azure VM. 😦 

So I guess this means we do try to get tenantID from the `AZURE_TENANT_ID` env var and if we come up empty-handed because the controller doesn't have its own managed identity, then we fall back on a value provided at install-time -- which we will document as being used only when tenantID cannot be inferred.

Reply:
Oh... you were already calling `azure.EnvironmentFromName(envName)`.